### PR TITLE
Add support for using doubles in the LUA53 build.

### LINKED ISSUE #3224: Add support for using doubles in LUA53 build
In order to improve compatibility with the LUA51 build, it would be very useful to have support for doubles in LUA53. Doubles are essential for storing times since the epoch if you want better than 1 second precision.

### PR DESCRIPTION
**Pull request by pjsg**
Fixes #3224.

- [x] This PR is for the `dev` branch rather than for `master`.
- [x] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [x] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/*`.

This brings back support for using doubles as the floating point format in the LUA53 build. This makes it much more compatible with the old LUA51 firmware.

### CODE REVIEW

File: docs/build.md [hunk 1]
```diff
@@ -76,8 +76,27 @@ editing `BIT_RATE_DEFAULT` in `app/include/user_config.h`:
...
+```c
+#define LUA_NUMBER_DOUBLE
+```
+
+OR by overriding this with the `make` command as it's [done during the CI
+build](https://github.com/nodemcu/nodemcu-firmware/blob/master/.travis.yml#L30):
+
+```
+make EXTRA_CCFLAGS="-DLUA_NUMBER_INTEGRAL ....
+```
```
Comment (HHHartmann) [inline docs/build.md:96]: cut and pasted too much here. Should be LUA_NUMBER_DOUBLE and might not be used in CI Build

File: docs/compiling.md [hunk 2]
```diff
@@ -61,7 +61,8 @@ just the `luac.cross` image.  You must first ensure that the following option in
 `app/include/user_config.h` is matched to your target configuration:
 
 ```c
-//#define LUA_NUMBER_INTEGRAL       // uncomment if you want an integer build
+//#define LUA_NUMBER_INTEGRAL       // uncomment if you want an integer build (LUA5.1)
```
Comment (HHHartmann) [inline docs/compiling.md:64]: maybe use Lua 5.1 or LUA51 but not a mixture of both.
Might also make sense to align these lines with the actual content of `app/include/user_config.h`

File: docs/build.md [hunk 3]
```diff
@@ -76,8 +76,27 @@ editing `BIT_RATE_DEFAULT` in `app/include/user_config.h`:
...
+
+### Integer build (LUA 5.1 only)
```
Comment (HHHartmann) [inline docs/build.md:98]:
```suggestion
### Integer build (Lua 5.1 only)
```

File: docs/build.md [hunk 4]
```diff
@@ -76,8 +76,27 @@ editing `BIT_RATE_DEFAULT` in `app/include/user_config.h`:
 Note that, by default, the firmware runs an auto-baudrate detection algorithm so that typing a few characters at boot time will cause
 the firmware to lock onto that baud rate (between 1200 and 230400).
 
-### Integer build
-By default a build will be generated supporting floating-point variables.
+### Double build (LUA 5.3 only)
```
Comment (HHHartmann) [inline docs/build.md:79]:
```suggestion
### Double build (Lua 5.3 only)
```

File: app/lua/luaconf.h [hunk 5]
```diff
@@ -181,6 +181,10 @@
   #endif // #if !defined LUA_INTEGRAL_LONGLONG
 #endif // #if !defined LUA_NUMBER_INTEGRAL
 
+#ifdef LUA_NUMBER_DOUBLE
+#error All LUA 5.1 builds use doubles. Do not specify this option.
```
Comment (HHHartmann) [inline app/lua/luaconf.h:185]:
```suggestion
#error All Lua 5.1 builds use doubles. Do not specify this option.
```

File: docs/build.md [hunk 6]
```diff
@@ -76,8 +76,27 @@ editing `BIT_RATE_DEFAULT` in `app/include/user_config.h`:
...
+By default a build will be generated supporting floating-point variables (floats) and integers.
+To increase the precision of the floating point variables, a double build can be created. This
+is also the default in the LUA 5.1 builds. The downside is that more memory is consumed when
```
Comment (HHHartmann) [inline docs/build.md:82]:
```suggestion
is also the default in the Lua 5.1 builds. The downside is that more memory is consumed when
```

**Comment by TerryE** [comment]
See comments on #3224.  The Lua VM evaluates integer expressions using the default (`sint32`) type.  Adding doubles alone will work in a very counter intuitive way for developer familiar with the 5.1 VM behaviour.

File: app/include/user_config.h [hunk 7]
```diff
@@ -42,21 +42,34 @@
...
+
+// When using LUA53, two different builds are now supported. 
+// The main difference is in the // processing of numeric data types.
+// If LUA_NUMBER_DOUBLE is defined, then doubles are used to hold floating
+// point numbers. Integers can be converted to doubles and back without any
+// loss of precision.
+// Otherwise all floating point operations use floats. Only integers under 2^24
+// can be represented exactly in floating point.
+// Note that LUA53 also supports Integers natively, but you have to be careful 
+// not to promote an integer to a float point variable if you are using a float build
```
[the remark this answers is not in the document]
Comment (pjsg) [reply app/include/user_config.h]: Done

File: app/lua53/luaconf.h [hunk 8]
```diff
@@ -86,8 +86,16 @@
 #endif
 #endif
 #  define LUA_INT_TYPE    LUA_INT_INT
+#ifdef LUA_NUMBER_DOUBLE
+#  define LUA_FLOAT_TYPE  LUA_FLOAT_DOUBLE
+#else
 #  define LUA_FLOAT_TYPE  LUA_FLOAT_FLOAT
+#endif
 //#  define LUA_FLOAT_TYPE  LUA_FLOAT_DOUBLE
```
[the remark this answers is not in the document]
Comment (pjsg) [reply app/lua53/luaconf.h]: I wasn't sure quite which line you meant, but I got rid of the commented out line.

File: docs/build.md [hunk 9]
```diff
@@ -76,8 +76,27 @@ editing `BIT_RATE_DEFAULT` in `app/include/user_config.h`:
 Note that, by default, the firmware runs an auto-baudrate detection algorithm so that typing a few characters at boot time will cause
 the firmware to lock onto that baud rate (between 1200 and 230400).
 
-### Integer build
-By default a build will be generated supporting floating-point variables.
+### Double build (LUA 5.3 only)
+By default a build will be generated supporting floating-point variables (floats) and integers.
```
[the remark this answers is not in the document]
Comment (pjsg) [reply docs/build.md]: Fixed

File: docs/build.md (hunk 3 above)
Comment (pjsg) [reply docs/build.md]: I fixed everything so that it was of the form "Lua 5.1" or "Lua 5.3"

File: docs/compiling.md (hunk 2 above)
Comment (pjsg) [reply docs/compiling.md]: Done

File: app/include/user_config.h [hunk 10]
```diff
@@ -42,21 +42,34 @@
...
-// operations, and decimal fraction constants are illegal.  Otherwise all
-// numeric operations use floating point, though they are exact for integer
-// expressions < 2^53.
-
-// The main advantage of INTEGRAL builds is that the basic internal storage unit,
-// the TValue, is 8 bytes long.  We have now reduced the size of FP TValues to
-// 12 bytes rather than the previous 16 as this gives a material RAM saving with
-// no performance loss.  However, you can define LUA_DWORD_ALIGNED_TVALUES and
-// this will force 16 byte TValues on FP builds.
+// operations, and decimal fraction constants are illegal.
```
[the remark this answers is not in the document]
Comment (pjsg) [reply app/include/user_config.h]: Done

**Comment by pjsg** [comment]
This version changes the option for a build to be double & 64-bit ints. In retrospect, I should have done this first time around.

File: docs/build.md (hunk 1 above)
Comment (pjsg) [reply docs/build.md]: Fixed.

**Comment by pjsg** [comment]
@HHHartmann I think that I have addressed all your issues.

**Comment by marcelstoer** [comment]
@nwf @pjsg as discussed, ok for you if we land this immediately after the next master drop (~1 week from now)?

**Comment by pjsg** [comment]
Sounds good to me.

**Pull request merged**